File: Merge-two-sorted-lists/mergeTwoSortedLists.java
public class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
                val = x;
        }

        public static ListNode createList(int... values) {
                ListNode head = new ListNode(values[0]);
                ListNode prev = head;
                for (int i = 1; i < values.length; i++) {
                        prev.next = new ListNode(values[i]);
                        prev = prev.next;
                }
                return head;
        }

        public ListNode filter(Predicate<ListNode> f) {

                ListNode node = this;
                while (node != null && !f.test(node)) {
                        node = node.next;
                }
                return node;
        }

        @Override
        public String toString() {

                return val + "->" + next;
        }

        @Override
        public boolean equals(Object o) {
                if (this == o)
                        return true;
                if (o == null || getClass() != o.getClass())
                        return false;
                ListNode listNode = (ListNode) o;
                return val == listNode.val &&
                                Objects.equals(next, listNode.next);
        }

        @Override
        public int hashCode() {
                return Objects.hash(val, next);
        }
}

public class Main {
        static ListNode mergeTwoLists(ListNode list1, ListNode list2) {

        }

        public static void main(String[] args) {
                ListNode list1 = ListNode.createList(1, 2, 4);
                ListNode list2 = ListNode.createList(1, 3, 4);
                ListNode result = mergeTwoLists(list1, list2);
        };

}

File: neetcode/merge-two-sorted-lists/merge-two-sorted-lists.ts
// Time o(n)
// Space o(n)

class ListNode {
    val: number
    next: ListNode | null
    constructor(val = 0, next = null) {
        this.val = val;
        this.next = next;
    }
}

function mergeTwoLists(list1, list2) {
    const dummyNode = new ListNode()
    let node: ListNode = dummyNode

    while (list1 && list2) {
        if (list1.val < list2.val) {
            node.next = list1
            list1 = list1.next
        } else {
            node.next = list2
            list2 = list2.next
        }
        node = node.next as ListNode
    }
    if (list1) {
        node.next = list1
    } else {
        node.next = list2
    }
    return dummyNode.next
}